### PR TITLE
pmdabpf: Add CO-RE execsnoop tool

### PR DESCRIPTION
**Pull request by smalinux**

### CODE REVIEW

File: src/pmdas/bpf/modules/GNUmakefile
Comment (smalinux) [inline src/pmdas/bpf/modules/GNUmakefile:68]: Is there a better way? I think so.

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [inline src/pmdas/bpf/modules/execsnoop.c:198]: *handle_args

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [inline src/pmdas/bpf/modules/execsnoop.c:301]: I do not know when to use this function!

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [inline src/pmdas/bpf/modules/execsnoop.c:134]: `meterics[1]`: Use enum instead of hardcoded numbers.

File: src/pmdas/bpf/modules/module.h
Comment (andreasgerstmayr) [inline src/pmdas/bpf/modules/module.h:156]: This line allocates 100 bytes, but in the next line the pointer to these 100 bytes gets overridden.
These 100 bytes in memory will be lost forever, because there is no pointer to this memory block anymore (resulting in a memory leak). Afaics this line should just be deleted, in the next line the `string` variable is assigned to it, which contains memory allocated by the `asprintf` function.

**Review by andreasgerstmayr — CHANGES_REQUESTED** [review]
Hi @smalinux, thanks for this PR!

Overall the changes look quite good, I've added a few comments regarding memory allocations and other stuff. Did you get a chance to run this module with valgrind (`sudo valgrind pmval -L -K clear -K add,157,/var/lib/pcp/pmdas/bpf/pmda_bpf,bpf_init bpf.execsnoop.comm`)?

File: src/pmdas/bpf/modules/execsnoop.c
Comment (andreasgerstmayr) [inline src/pmdas/bpf/modules/execsnoop.c:299]: This function blocks for `PERF_POLL_TIMEOUT_MS` and then returns the number of perf events of the queue it processed. It can return 0 if no new events were found. In our case, a new event is the execution of a new binary (`execve` syscall).

There is a perf event queue and we have a fixed maximum number of instances in PCP (i.e. we don't want one instance for every single `execve` syscall, that could accumulate to tens of thousands over time). In the bcc PMDA we solved this by having a `cache` queue with a maximum size of `process_count`, which holds the most recent perf events (first in first out), and a background thread which polls the perf event queue and adds new events to the `cache` queue. Then, the PMDA returns items from this queue when asked by the `fetch` callback.

@natoscott, @jasonk000 what approach do you recommend for this PMDA?
I'd suggest to poll for new perf events with a short timeout (I'd set it to less than 100ms, imho 100ms is already too long to block in a callback) in the `refresh` function, and store the events in a C array of the size `process_count`, where each value of this array is one event, and `execsnoop_fetch_to_atom` returns values from `cache_array[inst][item]`. When there are more than `process_count` events processed, it should start overwriting the oldest instance (event) of the `cache_array`.

@smalinux: I think this tool is one of the most difficult libbpf tools to port over to PCP. Quite a challenge for the first module! :100:

File: src/pmdas/bpf/modules/errno_helpers.c
Comment (smalinux) [reply src/pmdas/bpf/modules/errno_helpers.c:2]: > // Copied from the iovisor/bcc project at https://github.com/iovisor/bcc/blob/5f2a64ec101b2916558bb934c436d4a6d4ed157c/libbpf-tools/errno_helpers.c

I think it will be useful to add this for the commit message also...

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c:299]: @andreasgerstmayr Yes, I believe that once I have execsnoop, porting many other tools would be a piece of cake! :)

File: src/pmdas/bpf/modules/module.h
Comment (smalinux) [reply src/pmdas/bpf/modules/module.h:156]: Updated.

File: src/pmdas/bpf/modules/module.h
Comment (smalinux) [reply src/pmdas/bpf/modules/module.h:151]: Updated.

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c]: @andreasgerstmayr Done. Thank you.

Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c]: @andreasgerstmayr Done. Thanks.

**Comment by jasonk000** [comment]
This is looking really good @smalinux, great work. I have left a few comments for extra safety!

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c:416]: I think this will never happen because I have a fixed size buffer with max = `env.process_count` elements (this is max number of events I need) and always save the count of elements of this buffer into a global variable `queuelength` (this is the actual number of events.)
if `queuelength` > `env.process_count`, the oldest element will drop & free from the list.
and I start overwriting this buffer all the time with the latest events, 
This seems like a circular buffer,
and I reset the head whole point to the first element of that buffer every `_refresh`.
and I loop over this buffer while through `_fetch_to_atom`

Also, I did a very simple a draft before starting this If you would like to check the logic behind that:
https://github.com/smalinux/tailq-example
`$ clone & make & ./tailq_ex`

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c]: updated.

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c]: updated. Thanks.

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c]: Updated.

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c:270]: Updated.

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c]: Nice Note, Just added one commit for this, I added `bpf.execsnoop.lost`  in a separate InDom has one instance represent the count of lost events.
Expected output: ( because I don't know how I could make my system busy enough to drop some events!)
```
$ pminfo -f bpf.execsnoop.lost

bpf.execsnoop.lost
    inst [0 or "0"] value 0
```
I know. It's not that much perfect, I think It will be better if I did something like that:
multiple instance each one represents one CPU (hence external instance name)
```
$ pminfo -f bpf.execsnoop.lost

bpf.execsnoop.lost
    inst [0 or "CPU 0"] value 10
    inst [0 or "CPU 10"] value 4
    inst [0 or "CPU 100"] value 20
```
but since I didn't test that even with iovisor/execsnoop, I couldn't handle that,
It seems not a big deal but It will need surgery ^^"
Maybe I could handle this later after merging this PR...

Thanks : )

File: src/pmdas/bpf/modules/module.h
Comment (smalinux) [reply src/pmdas/bpf/modules/module.h]: Updated. It's become so simple I don't know why I did it like that. Thanks!

File: src/pmdas/bpf/modules/module.h
Comment (smalinux) [reply src/pmdas/bpf/modules/module.h:153]: I got a warning from gcc:
```
module.h:152:9: warning: ignoring return value of ‘asprintf’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  152 |         asprintf(&string, "%d", i);                                                               
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~ 
```
Also, I did that inspired by this line [[Link](https://github.com/performancecopilot/pcp/blob/47093257c1ec139b4c36030cad6ba7a8deef1aca/src/pmdas/bpf/modules/module.h#L137)] ...

File: src/pmdas/bpf/modules/syscall_helpers.c
Comment (smalinux) [reply src/pmdas/bpf/modules/syscall_helpers.c:1]: | I assume these are completely unmodified from upstream?
Yes, If I go in this direction I think I have to clone the whole `iovisor/bcc` repo under `vendor/` like the reset other repositories under this dir.
Are you okay with that? @jasonk000 @natoscott

File: src/pmdas/bpf/modules/execsnoop.c
Comment (jasonk000) [reply src/pmdas/bpf/modules/execsnoop.c:270]: i still see `strcpy` here? maybe github is confused

File: src/pmdas/bpf/modules/execsnoop.c
Comment (jasonk000) [reply src/pmdas/bpf/modules/execsnoop.c:416]: I was more concerned that this assumes `fetch_to_atom` is called in a specific order. I'm not sure if that is guaranteed to hold in all scenarios as part of the API contract. As per the contract `inst` argument should be used to look up the correct instance in the tailq for each fetch:

```
For each instance listed in the profile (see [pmdaprofile](https://linux.die.net/man/3/pmdaprofile)(3)) of each metric listed in pmidlist, the pmdaFetchCallBack method is called to fill the pmAtomValue structure identified by avp with a value for a specific metric-instance pair identified by the metric descriptor mdesc and the instance inst. This value is then copied into the pmResult structure.
```
https://linux.die.net/man/3/pmdasetfetchcallback

File: src/pmdas/bpf/modules/execsnoop.c
Comment (smalinux) [reply src/pmdas/bpf/modules/execsnoop.c]: oh sorry, should be updated now

File: src/pmdas/bpf/modules/syscall_helpers.c
Comment (jasonk000) [reply src/pmdas/bpf/modules/syscall_helpers.c:1]: I would defer to @natoscott or @andreasgerstmayr

Comment (natoscott) [reply src/pmdas/bpf/modules/syscall_helpers.c:1]: Yeah, hmm, that's alot of relatively fast-moving code to be vendoring - I'm hesitant to do so.  Could we vendor just the bcc content below bcc/libbpf-tools or do we need more than that?  (is a subdir even possible with git-subtree @andreasgerstmayr , I wonder?)  That'd be more in keeping with the kind of thing we tend to vendor into PCP (small libraries for the most part).

File: src/pmdas/bpf/modules/execsnoop.c
Comment (natoscott) [reply src/pmdas/bpf/modules/execsnoop.c:411]: From memory I don't think there's a guarantee of any specific inst ordering for the fetch routine.  It'd be worth checking the fetch code below src/libpcp_pmda/src though.

Another option may be to provide a custom fetch routine instead of pmdaFetch if that helps, then you could take control over how the instances are traversed - an example of a PMDA doing this is src/pmdas/sample/src/sample.c

File: src/pmdas/bpf/modules/errno_helpers.c
Comment (natoscott) [reply src/pmdas/bpf/modules/errno_helpers.c:2]: Sorry, missed this question earlier - @andreasgerstmayr as mentioned below, I wonder if we can vendor the libbpf-tools directory?

File: src/pmdas/bpf/modules/syscall_helpers.c
Comment (smalinux) [reply src/pmdas/bpf/modules/syscall_helpers.c:1]: | We only need the *_helpers.*

`*_helpers.*` and `*.bpf.c`

Comment (smalinux) [reply src/pmdas/bpf/modules/syscall_helpers.c:1]: ah, I see, thanks a lot, I suggest adding something like this to GNUMakefile: 
`$ cp vendor/bpf-tools/*.bpf.ch  pmda/bpf/modules/`

Comment (andreasgerstmayr) [reply src/pmdas/bpf/modules/syscall_helpers.c:1]: I suggest using symbolic links, for example: https://github.com/performancecopilot/pcp/blob/07dc727f9eb5749669dd2ea213b7d360d2554b63/src/libpcp_web/src/GNUmakefile#L121-L122

**Review by andreasgerstmayr — APPROVED** [review]

**Comment by andreasgerstmayr** [comment]
Thank you very much for your contribution!

**Pull request merged**
